Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54328 Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants**

Differential Revision: [D27369134](https://our.internmc.facebook.com/intern/diff/D27369134)

anjali411 committed:
Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj"

[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj""




[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj""




[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj""




[ghstack-poisoned]

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -130,3 +130,33 @@ def fn(x: complex):
             self.checkScript(fn, (val, ))
+
+    def test_cmath_constants(self):
+        def checkCmathConst(const_name, ret_type="float"):
```
```suggestion
        def checkCmathConst(const_name, ret_type):
```

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -130,3 +130,33 @@ def fn(x: complex):
+        float_consts = ['pi', 'e', 'tau', 'inf', 'nan']
+        complex_consts = ['infj', 'nanj']
+        for x in float_consts:
+            checkCmathConst(x)
```
```suggestion
            checkCmathConst(x, "float")
```

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -130,3 +130,33 @@ def fn(x: complex):
+                    .format(const_name=const_name, res_python=res_python, res_script=res_script))
+                self.assertEqual(res_python, res_script, msg=msg)
+
+        float_consts = ['pi', 'e', 'tau', 'inf', 'nan']
+        complex_consts = ['infj', 'nanj']
```
Is there any way to define these using `cmath` so that new constants that are added get tested automatically?

Unlikely scenario but I thought I'd ask.

SplitInfinity reviewed (commented):
Everything looks good but some tests are failing. I strongly suspect that they are inherited from #54089 but I will hold off on stamping until we are sure.

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants"




[ghstack-poisoned]

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -109,3 +109,30 @@ def fn(x: complex):
+
+    def test_cmath_constants(self):
+        def checkCmathConst(const_name):
+            funcs_template = dedent('''
+                def func():
+                    return cmath.{const}
+                ''')
+
+            funcs_str = funcs_template.format(const=const_name)
+            scope = {}
+            execWrapper(funcs_str, globals(), scope)
+            cu = torch.jit.CompilationUnit(funcs_str)
+            f_script = cu.func
+            f = scope['func']
+
+            res_python = f()
+            res_script = f_script()
+
+            if res_python != res_script:
+                msg = ("Failed on {const_name}. Python: {res_python}, Script: {res_script}"
+                       .format(const_name=const_name, res_python=res_python, res_script=res_script))
+                self.assertEqual(res_python, res_script, msg=msg)
```
Looks similar to the testing code from the PR below, consider doing some refactoring.

anjali411 replied:
good idea! done

SplitInfinity commented on torch/csrc/jit/serialization/import_source.cpp:
```diff
@@ -222,6 +222,16 @@ struct SourceImporterImpl : public Resolver,
       return std::make_shared<SimpleValue>(
           graph->insertConstant(std::numeric_limits<double>::quiet_NaN(), loc));
     }
+    if (name == "infj") {
+      return std::make_shared<SimpleValue>(graph->insertConstant(
+          c10::complex<double>(0, std::numeric_limits<double>::infinity()),
+          loc));
+    }
+    if (name == "nanj") {
+      return std::make_shared<SimpleValue>(graph->insertConstant(
+          c10::complex<double>(0, std::numeric_limits<double>::quiet_NaN()),
+          loc));
+    }
```
It seems weird to me that these constants would be present in the serialized code as `infj` and `nanj` without any qualifying namespace or module like `cmath`. Is that because of how complex numbers are printed (i.e. `[imaginary part]j`)? Can you include a snippet of serialized code that contains complex constants (just for my own education)?

anjali411 replied:
Python Code:
```
In [12]: def fn():
    ...:     return cmath.infj
    ...:

In [13]: torch.jit.save(torch.jit.script(fn), "infj.pt")
```

Serialized code:
```
class PlaceholderModule(Module):
  __parameters__ = []
  __buffers__ = []
  training : bool
  def forward(self: __torch__.PlaceholderModule) -> complex:
    return 0.+infj
```

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants"


Differential Revision: [D27369134](https://our.internmc.facebook.com/intern/diff/D27369134)

[ghstack-poisoned]

anjali411 committed:
Update on "Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants"


Differential Revision: [D27369134](https://our.internmc.facebook.com/intern/diff/D27369134)

[ghstack-poisoned]
